### DISCUSSION TITLE
How to modify a module content in the processAssets hook!

### DISCUSSION THREAD

**Discussion by BUPTlhuanyu** [webpack/webpack#15719] [Q&A]
I hava some work to collect information from some less files, but how can i modify a module content in the processAssets hook!
```
import data from 'virtual.modulation';

data.get('bundle_a');
```

**Comment by vankop** [comment]
you can create a loader for that

**Reply by BUPTlhuanyu** [reply]
But i got an error
```
Module not found: Error: Can't resolve 'virtual.modulation'
```
```
  module: {
    rules: [
      {
        test: /virtual.modulation$/,
        use: [function fn(content) {
          return 'module.exports = global.${globaleModuleName};';
        }],
      },
    ],
  },
```

**Reply by BUPTlhuanyu** [reply]
I use the webpack-virtual-modules to handle the error of module not found, but how can i modify a module content in the processAssets hook?

**Reply by alexander-akait** [reply]
You can't do it, you need loader here

**Reply by BUPTlhuanyu** [reply]
Why?
But how to handle the error that module not found here? The module is a virtual module, it is not exist.

**Reply by alexander-akait** [reply]
By design, `processAssets` for assets, you have a module (modules will be bundled in chunks, from chunks we will create JS assets and they will be emitted), also webpack handle dependecies (i.e. `import`/`require`/etc) before `processAssets`, so you need to have a loader to handle it

**Reply by BUPTlhuanyu** [reply]
step1: i use webpack-virtual-modules to handle the error that virtual module not found
step2: i use a loader to match the module name，and change the content of it.
step3: i change the assets source by string.prototype.replace
```
const path = require('path');
const VirtualModulesPlugin = require('webpack-virtual-modules');

function VirtualModule() {}

VirtualModule.prototype.apply = function (compiler) {
  const modulationPath = path.resolve(process.cwd(), './node_modules/virtual.modulation');
  const virtualModules = new VirtualModulesPlugin({ [modulationPath]: `module.exports = '';` });
  virtualModules.apply(compiler);

  compiler.hooks.compilation.tap('VirtualModule', function (compilation) {
    compilation.hooks.processAssets.tap('replace', function () {
      console.log('compilation', compilation);
      let code = compilation.assets['main.js'].source().toString().replace('REPLACE_ME START', compilation.coleectedData);
      compilation.assets['main.js'] = new WebpackSources.RawSource(code);
    });
  });
};

module.exports = {
  entry: './index.js',
  output: {
    libraryTarget: 'umd',
  },
  plugins: [new VirtualModule()],
  module: {
    rules: [
      {
        test: /virtual.modulation$/,
        use: [
          {
            loader: path.resolve(__dirname, './loader.js'),
          },
        ],
      },
    ],
  },
};
```
./loader.js
```
module.exports = function () {
  return `
    var data = 'REPLACE_ME START';
    module.exports = data;
  `;
};

```

**Reply by BUPTlhuanyu** [reply]
it seems pretty hacky to replace the content of a module, but i have no idea to handle this properly!

**Reply by HomyeeKing** [reply]
+1, very unnatural compare to rollup
